Feature #39363: Add Advanced Tag search functionality to DAGs list

diogotrodrigues opened the pull request:
### Motivation:
Improve the user experience by enabling support for Advanced Tag search using logical operators AND, OR and ( ) for precedence. This allows users to construct complex queries based on tags, making it significantly easier to locate specific DAGs in large or complex environments.

---

### Example of usage:

Supports complex queries with nested logic, such as "(tag1 OR tag2) AND tag3" or even
 "((tag1 OR tag2) AND tag3) OR tag4."

When you write something like tag1 OR tag2 AND tag3, it behaves as you'd intuitively expect: the AND condition is evaluated first, followed by the OR.

***Note:*** The AND and OR Operators are case-insensitive.

---

### What was done?:
* Added the logic to parse and handle advanced tag search queries;
* Modified the necessary API endpoints;
* Added tests to ensure the functionality works as expected;
* Added UI components to support advanced tag search.

---

### Here is a quick video demonstrating the feature!

https://github.com/user-attachments/assets/67cc69b9-fd6c-4085-8814-12b52112b063



<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #39363

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---

diogotrodrigues committed:
Feature #39363: Add Advanced Tag search functionality to DAGs list

Motivation:
Improve the user experience by enabling support for Advanced Tag
search using logical operators AND, OR and parentheses for
precedence. This allows users to construct complex queries
based on tags, making it significantly easier to locate specific
DAGs in large or complex environments.

What was done?:
Added the logic to parse and handle advanced tag search queries;
Modified the necessary API endpoints;
Added tests to ensure the functionality works as expected;
Added UI components to support advanced tag search.

Co-authored-by: João Botelho <joao.v.botelho@tecnico.ulisboa.pt>

pierrejeambrun reviewed (commented):
Great, nice PR!

Some CI checks are failing, you need to add `pyparsing` to the project dependencies if it's not already there.

Also you can run `yarn lint` or `pre-commit run ts-compile-format-lint-ui --all-files` to run linting on the UI. (Which is currently failing).

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/DagsList/DagsFilters/AdvancedSearch.tsx:
```diff
@@ -0,0 +1,122 @@
+  const { data: tagsData, isLoading } = useDagServiceGetDagTags({ orderBy: "name" });
+  const availableTags = tagsData?.tags ?? [];
+
+  useEffect(() => {
+    setInputValue(initialValue);
+  }, [initialValue, isOpen]);
```
Generally, we should not be updating state inside of a `useEffect`. I think it would be better to unmount this component when its closed and then we wouldn't need this useEffect

diogotrodrigues replied:
Thanks for the tip! we will apply that change !

pierrejeambrun replied:
That comment was resolved but the useEffect is still there, is that expected?

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/DagsList/DagsFilters/AdvancedSearch.tsx:
```diff
@@ -0,0 +1,122 @@
+                <Button colorScheme="gray" onClick={handleSubmit} variant="outline">
+                  {translate("common:enter")}
+                </Button>
+              </Box>
```
Let's put this on its own row at the bottom of the dialog. I was confused at first and didnt realize this was the submit for the whole search bar

diogotrodrigues replied:
like how we did on this new commit? or you had other idea on how to restructure it ?

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/DagsList/DagsList.tsx:
```diff
@@ -239,11 +260,17 @@ export const DagsList = () => {
       <VStack alignItems="none">
         <SearchBar
-          buttonProps={{ disabled: true }}
+          buttonProps={{ disabled: false, onClick: () => setShowAdvanced(true) }}
```
```suggestion
          buttonProps={{ onClick: () => setShowAdvanced(true) }}
```

bbovenzi commented on airflow-core/src/airflow/ui/src/constants/searchParams.ts:
```diff
@@ -17,6 +17,7 @@
  * under the License.
  */
 export enum SearchParamsKeys {
+  ADVANCED_TAGS = "advanced_tags",
```
I think it would be better if we used the existing tags and if we don't get any "and"/"or" options, we just assume the current behavior of "and".

Also, this prevents the user from getting confused by having two different tags filters in the UI

diogotrodrigues replied:
Could you please clarify this, I couldn't understand it. Thank you for your time and help.

bbovenzi reviewed (commented):
UX-wise I wonder if it would be better to look more like a dropdown from the existing search bar instead of opening a new dialog

<img width="621" alt="Screenshot 2025-06-12 at 1 55 46 PM" src="https://github.com/user-attachments/assets/b54b72aa-ff7c-400d-b2ad-563d55a6ded0" />

Johnny1cyber committed:
Refactor: added missing dependency and modified ui

Added pyparsing to airflow-core pyproject.toml;
Removed unecessary code;
Changed dialog close/enter buttons positions.

Co-authored-by: Diogo Rodrigues <diogo.teixeira.rodrigues@tecnico.ulisboa.pt>

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/DagsList/DagsFilters/AdvancedSearch.tsx:
```diff
@@ -0,0 +1,122 @@
+                />
+              </Box>
+            </Box>
+            <Box display="flex" gap={4} justifyContent="flex-end">
```
Let's put the clear and enter buttons inside of `<Dialog.Footer>` instead

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/DagsList/DagsFilters/AdvancedSearch.tsx:
```diff
@@ -0,0 +1,122 @@
+            <Box display="flex" flexDirection="row" gap={3}>
+              <Box minW="160px">
+                <ReactSelect
+                  aria-label={translate("list.selectTag", { defaultValue: "Add Tag" })}
```
Let's make sure to translate "Add Tag"

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/DagsList/DagsFilters/AdvancedSearch.tsx:
```diff
@@ -0,0 +1,122 @@
+                    handleSubmit();
+                  }
+                }}
+                placeholder={translate("Advanced tag query, e.g. (tag1 OR tag2) AND tag3")}
```
Let's actually add these keys to the `en/dags.json` file

pierrejeambrun replied:
need some work.

bbovenzi reviewed (commented):
A few quick comments. Let's still try to get this in for 3.1!

pierrejeambrun reviewed (commented):
Great to see activity on this PR.

CI need fixing, and feel free to close comments you have already addressed.

I believe there are a bunch of comment that still need to be addressed, ping me when the PR is ready for another round of review

diogotrodrigues committed:
Refactor Advanced Search

Updated AdvancedSearch.tsx to accommodate the requested changes.

Co-authored-by: Diogo Rodrigues <diogo.teixeira.rodrigues@tecnico.ulisboa.pt>

jason810496 commented on airflow-core/src/airflow/api_fastapi/common/parameters.py:
```diff
@@ -432,6 +442,76 @@ def depends(
+    def __init__(self, value: str | None = None, skip_none: bool = True):
+        super().__init__(value, skip_none)
+
+    def parse_advanced_query(self, query: str):
```
Is it possible to add type annotation for the return type?

jason810496 commented on airflow-core/src/airflow/api_fastapi/common/parameters.py:
```diff
@@ -432,6 +442,76 @@ def depends(
+            return parsed
+        except ParseException as e:
+            raise HTTPException(400, f"Invalid tag query: {e}")
+
+    def _expr_to_sqlalchemy(self, expr):
```
Missing here as well.

jason810496 commented on airflow-core/src/airflow/api_fastapi/common/parameters.py:
```diff
@@ -432,6 +442,76 @@ def depends(
+                    return and_(left_cond, right_cond)
+                if op.upper() == "OR":
+                    return or_(left_cond, right_cond)
+            # If the structure is not as expected, raise an error
+            raise HTTPException(400, "Invalid tag query structure")
+        raise HTTPException(400, "Invalid tag query structure")
```
Is it possible to make the error message more verbose? Maybe like having hint for missing on which side, etc.
Nice to have if it's possible, but not strong opinion for having it.

jason810496 reviewed (commented):
Thanks for the PR! Nice feature!

The static checks still need fix, thanks! 
We just switch `pre-commit` to `prek` recently, you install `prek` by `uv tool install prek`.
```bash
# for ruff 
prek ruff --all-files && prek run ruff-format --all-files

# for frontend lint
prek ts-compile-lint-ui --all-files
```

pierrejeambrun reviewed (commented):
@diogotrodrigues are you still working on that? It would be a great addition to `3.2.0`. 

There are still some comments unaddressed. Also please feel free to mark as `resolved` things that you solved, so it's easier for reviewer to see if works remains before we need to do the next review.
